src/lib/serviceRoutines/postQueryContext.cpp

```diff
@@ -241,26 +178,159 @@ std::string postQueryContext
+      //
+      if (aP->providingApplication == "")
+      {
+        if (aP->found == false)
+        {
+          continue;  // Non-found pairs of entity/attribute are thrown away
+        }
+
+        QueryContextResponse* qP = new QueryContextResponse(eP, aP);
+        responseV.push_back(qP);
+      }
+      else if (requestP == NULL)
+      {
+        requestP = new QueryContextRequest(aP->providingApplication, eP, aP->name);
+        requestV.push_back(requestP);
+        ++noOfForwards;
+      }
+      else
+      {
+        requestP->attributeList.push_back_if_absent(aP->name);
+        requestP->entityIdVector.push_back_if_absent(eP);
+      }
+    }
+  }
+
+  //
+  // If no redirectioning is necessary, just return the result
+  //
+  if (noOfForwards == 0)
+  {
+    //
+    // In case the response is empty, fill response with the entity we looked for
+    //
+    if (qcrsP->contextElementResponseVector.size() == 0)
+    {
+      //
+      // FIXME P6: What do we do if we have more than one entity in incoming entityIdVector?
+      //           Push each entityId in a separate contextElementResponse?
+      //           Also, perhaps mongoQueryContext should implement this and not postQueryContext
+      //           This FIXME is related to github issue #588 and (probably) #650.
+      //           Also, optimizing this would be part of issue #768
+      //
+
+
+      //
+      // It's better not to give any details here as we'd have to choose ONE of the entities in the vector ...
+      // std::string details = "Entity id: /" + parseDataP->qcr.res.entityIdVector[0]->id + "/";
+      //
+      qcrsP->errorCode.fill(SccContextElementNotFound);
+    }
+
+    answer = qcrsP->render(ciP, QueryContext, "");
+    return answer;
+  }
+
+
+  //
+  // DEBUG - present the vector prepared for Context Provider Forwarding
+  //
+  requestV.present();
+
+
+  //
+  // Now, send 'noOfForwards' Query requests, each in a separate thread and
+  // await all the responses.
+  // Actually, if there is only ONE forward to be done then there is no reason to
+  // do the forward in a separate shell. Better to do it inside the current thread.
+  //
+  // If providingApplication is empty then that part of the query has been performed already, locally.
+  // 
+  //
+  for (unsigned int fIx = 0; fIx < requestV.size(); ++fIx)
   {
-    qcrsP->contextElementResponseVector[0]->statusCode.details = std::string("Redirected to context provider ") + ip + ":" + portV + prefix;
+    if (requestV[fIx]->contextProvider == "")
+    {
+      //
+      // No forward, we already have the result in responseV
+      //
+      continue;
+    }
+
+    QueryContextResponse* qP = new QueryContextResponse();
+    queryForward(ciP, requestV[fIx], qP);
+    responseV.push_back(qP);
   }
 
-  answer = qcrsP->render(ciP, QueryContext, "");
+  responseV.present();
+
+  //
+  // Is qcrsAccumulatedP OK? Without doing anything special here ... ?
```

Comment:
What is qcrsAccumulatedP ? I don't find this variable... Maybe the comment is stale?